libraries/hotstuff/include/eosio/hotstuff/qc_chain.hpp

```diff
@@ -77,71 +77,261 @@ namespace eosio::hotstuff {
+
+   class pending_quorum_certificate {
    public:
-      explicit quorum_certificate(size_t finalizer_size = 0) {
-         active_finalizers.resize(finalizer_size);
+      enum class state_t {
+         unrestricted,  // No quorum reached yet, still possible to achieve any state.
+         restricted,    // Enough `weak` votes received to know it is impossible to reach the `strong` state.
+         weak_achieved, // Enough `weak` + `strong` votes for a valid `weak` QC, still possible to reach the `strong` state.
+         weak_final,    // Enough `weak` + `strong` votes for a valid `weak` QC, `strong` not possible anymore.
+         strong         // Enough `strong` votes to have a valid `strong` QC
+      };
+
+      struct votes_t {
+         hs_bitset     _bitset;
+         bls_signature _sig;
+
+         void resize(size_t num_finalizers) { _bitset.resize(num_finalizers); }
+         size_t count() const { return _bitset.count(); }
+
+         bool add_vote(const std::vector<uint8_t>& proposal_digest,
+                       size_t index,
+                       const bls_public_key& pubkey,
+                       const bls_signature& new_sig) {
+            if (_bitset[index])
+               return false; // shouldn't be already present
+            if (!fc::crypto::blslib::verify(pubkey, proposal_digest, new_sig))
+               return false; 
+            _bitset.set(index);
+            _sig = fc::crypto::blslib::aggregate({ _sig, new_sig }); // works even if _sig is default initialized (fp2::zero())
+            return true;
+         }
+
+         void reset(size_t num_finalizers) {
+            if (num_finalizers != _bitset.size())
+               _bitset.resize(num_finalizers);
+            _bitset.reset();
+            _sig = bls_signature();
+         }
+      };
+
+      pending_quorum_certificate() = default;
+
+      explicit pending_quorum_certificate(size_t num_finalizers, size_t quorum) :
+         _num_finalizers(num_finalizers),
+         _quorum(quorum) {
+         _weak_votes.resize(num_finalizers);
+         _strong_votes.resize(num_finalizers);
       }
 
-      explicit quorum_certificate(const quorum_certificate_message& msg, size_t finalizer_count)
-              : proposal_id(msg.proposal_id)
-              , active_finalizers(msg.active_finalizers.cbegin(), msg.active_finalizers.cend())
-              , active_agg_sig(msg.active_agg_sig) {
-               active_finalizers.resize(finalizer_count);
+      explicit pending_quorum_certificate(const fc::sha256& proposal_id,
+                                          const digest_type& proposal_digest,
+                                          size_t num_finalizers,
+                                          size_t quorum) :
+         pending_quorum_certificate(num_finalizers, quorum) {         
+         _proposal_id = proposal_id;
+         _proposal_digest.assign(proposal_digest.data(), proposal_digest.data() + 32);
+         _state = state_t::unrestricted;
```

Comment:
`_state` already is defaulted to unrestricted.

Reply:
removed